src/Protocols/DfLikeAlternative.hs

```diff
@@ -0,0 +1,416 @@
+    = Vec wdBytes (Maybe (BitVector 8))
+  type OtpPayload
+    (Reverse (Axi4WriteAddress dom 'KeepBurst waKeepSize lw iw aw waKeepRegion waKeepBurstLength waKeepLock waKeepCache waKeepPermissions waKeepQos waUser),
+     Reverse (Axi4WriteData dom 'KeepStrobe wdBytes wdUser),
+     Axi4WriteResponse dom 'KeepResponse iw wrUser)
+    = ()
+  type DfLikeParam
+    (Reverse (Axi4WriteAddress dom 'KeepBurst waKeepSize lw iw aw waKeepRegion waKeepBurstLength waKeepLock waKeepCache waKeepPermissions waKeepQos waUser),
+     Reverse (Axi4WriteData dom 'KeepStrobe wdBytes wdUser),
+     Axi4WriteResponse dom 'KeepResponse iw wrUser)
+    = (BitVector (Width aw), wrUser)
+    -- write data address, user data for write response
+
+  dfLikeCkt (_, param) = dfLikeCktHelper s0 blankOtp (stateFn param) where
+    s0 = (Nothing, S2M_NoWriteResponse)
+   
+    blankOtp = (S2M_WriteAddress{_awready = False}, S2M_WriteData{_wready = False}, S2M_NoWriteResponse)
+   
+    stateFn (dataAddr,wrUser) (wAddrVal, wDataVal, wRespAck) ack _ = do
+      wAddrAck <- processWAddr wAddrVal
+      (wDataAck, inpItem) <- processWData wDataVal
+      wRespVal <- gets snd
+      processWRespAck
+      pure ((wAddrAck, wDataAck, wRespVal), inpItem, False)
+      where
+   
+      processWAddr M2S_NoWriteAddress = pure (S2M_WriteAddress{_awready = False})
+      processWAddr M2S_WriteAddress{ _awburst } | _awburst /= BmFixed = pure (S2M_WriteAddress{_awready = True})
+      processWAddr M2S_WriteAddress{_awaddr, _awid} = do
+        (_,b) <- get
+        put (if _awaddr == dataAddr then Just _awid else Nothing, b)
+        pure (S2M_WriteAddress{_awready = True})
+   
+      processWData M2S_NoWriteData = pure (S2M_WriteData{_wready = False}, Nothing)
+      processWData M2S_WriteData{_wlast, _wdata} = do
+        (shouldRead,respS2M) <- get
+        -- we only want to output _wready = false if we're the recpient of the writes AND ack is false
+        -- we only want to return data if we're the recpient of the writes
+        -- we only want to output on writeresponse if we're the recipient of the writes AND ack is true
+        case (shouldRead, ack) of
+          (Nothing, _) -> pure (S2M_WriteData{_wready = True}, Nothing)
+          (Just _, False) -> pure (S2M_WriteData{_wready = False}, Just _wdata)
+          (Just sr, True) -> do
+            put (Nothing,
+                 if _wlast then S2M_WriteResponse {_bid = sr, _bresp = ROkay, _buser = wrUser } else respS2M)
+            pure (S2M_WriteData{_wready = True}, Just _wdata)
+   
+      processWRespAck = when (_bready wRespAck) $ modify (\(a,_) -> (a,S2M_NoWriteResponse))
+
+-- Always sends 'ROkay' along 'WriteResponse' channel
+instance (NFDataX dat, NFDataX rdUser, KnownNat (Width aw), KnownNat (Width iw)) =>
+  DfLikeAlternative
+    (Reverse (Axi4ReadAddress dom 'KeepBurst 'NoSize lw iw aw keepRegion 'KeepBurstLength keepLock keepCache keepPermissions keepQos raData),
```

Comment:
I think I've asked this before, but I forgot: why are some of these properties hardcoded, and some aren't?

We should probably move this to a single type argument like you did in another PR, btw.

Reply:
The hardcoded properties are the ones that `dfLikeCircuit` looks at, since there's currently no support for "looking, but with a default value if the field isn't there". That other PR adds support for looking with a default value, so once that's added we can get rid of the hardcoded properties.